The pipeline and scripts here constitute a secondary analysis pipeline based on the output of the primary pipeline circleseq. This pipeline takes a genome annotation file (gff) and an uncompressed vcf header representing the species of interest (can be generated by coercing to vcf output with a bam from the primary pipeline and copying the header), along with a pileup (variants.txt) from the circleseq pipeline output to evaluate the distribution and effect of somatic mutations across individual genes and entire gene ontology sets.

Has additional external dependency SnpEff (http://snpeff.sourceforge.net/) and the python package GFFUtils (0.10 or greater). 

To use, first edit the Snakefile config section (at the top) to reflect your system layout, then type into the shell:
snakemake {sample_name}_genemut.txt

Where sample name is the prefix for the variants file (e.g. {sample_name}_variants.txt). 

Graphs can be produced by calling "python3 visualize_synonymity.py" on the resulting custom output text file. 
